cmd/yesiscan/main.go

```diff
@@ -95,301 +113,398 @@ func CLI(program, version string, debug bool, logf func(format string, v ...inte
-				s := c.Args().Get(i)
-				args = append(args, s)
+		}
+		if config.Backends != nil {
+			for k, v := range config.Backends {
+				backends[k] = v // copy
 			}
+		}
+	}
 
-			// is there at least one yes-?
-			isAdditive := false
-			for _, f := range lib.Backends {
-				if c.Bool(fmt.Sprintf("yes-backend-%s", f)) {
-					isAdditive = true
-				}
+	// Command line options override anything in the config.
+	if c.IsSet("auto-config-uri") {
+		autoConfigURI = c.String("auto-config-uri")
+	}
+	if c.IsSet("auto-config-cookie-path") {
+		autoConfigCookiePath = c.String("auto-config-cookie-path")
+	}
+	if c.IsSet("quiet") {
+		quiet = c.Bool("quiet")
+	}
+	if c.IsSet("regexp-path") {
+		regexpPath = c.String("regexp-path")
+	}
+	// config-path makes no sense here
+	if c.IsSet("output-type") {
+		outputType = c.String("output-type")
+	}
+	if c.IsSet("output-path") {
+		outputPath = c.String("output-path")
+	}
+	if c.IsSet("output-s3bucket") {
+		outputS3Bucket = c.String("output-s3bucket")
+	}
+	if c.IsSet("region") {
+		region = c.String("region")
+	}
+	if c.IsSet("profile") {
+		profiles = []string{} // erase any previous
+		for _, x := range c.StringSlice("profile") {
+			profiles = append(profiles, x)
+		}
+	}
+
+	// auto config URI magic...
+	if autoConfigURI != "" { // we must try to auto config
+		logf("getting config from: %s", autoConfigURI)
+		data, err := DownloadConfig(autoConfigURI)
+		if err != nil {
+			return errwrap.Wrapf(err, "autoConfigURI download failed on: %s", autoConfigURI)
+		}
+		p, err := GetConfigPath(c.String("config-path"))
+		if err != nil {
+			return err
+		}
+		b, err := os.ReadFile(p)
+		if os.IsNotExist(err) { // no config exists here...
+			b = []byte{} // (implied, but now cleanly initialized)
+			err = nil    // clear this
+		} else if err != nil {
+			return err
+		}
+
+		isJson := func(d []byte) error {
+			buffer := bytes.NewBuffer(d)
+			if buffer.Len() == 0 {
+				return fmt.Errorf("empty config file")
 			}
+			decoder := json.NewDecoder(buffer)
 
-			// isBackendEnabled specifies if a particular backend
-			// should be enabled based on the lookup of the flags.
-			isBackendEnabled := func(f string) bool {
-				if isAdditive && c.Bool(fmt.Sprintf("yes-backend-%s", f)) {
-					return true
-				}
+			var configData Config // this gets populated during decode
+			err := decoder.Decode(&configData)
+			return errwrap.Wrapf(err, "invalid json")
+		}
 
-				if !isAdditive && !c.Bool(fmt.Sprintf("no-backend-%s", f)) {
-					return true
-				}
+		// if equal, we don't need to change the config...
+		// check it's valid json before writing it? (for portal errors)
+		if err := isJson(data); !bytes.Equal(data, b) && err == nil {
 
-				return false
+			// store new config file
+			logf("writing new config...")
+			// XXX: set umask for u=rw,go=
+			if err := os.WriteFile(p, data, interfaces.Umask); err != nil {
+				return errwrap.Wrapf(err, "autoConfigURI store failed on: %s", p)
 			}
```

Comment:
Overwriting a config file with one downloaded from a URI violates the principal of least surprise. Users generally expect configs to be read-only, and may have settings there that would unexpectedly and irretrievably be lost, including the `auto-config-uri` setting that got you here.

Reply:
Already discussed above. But it does not violate this principle since it would only ever happen if someone turned this option on to begin with. IOW, don't enable auto-steering if you don't want your car to steer. 99% of users may never see or touch this. The enterprise users that do will be glad it's actually fully automatic. That's the point of it =D

Reply:
Apologies for unresolving this thread. I've solicited other opinions on the topic.